### PR TITLE
fix: Incorrect up-mapping for tokens in derive attributes

### LINKED ISSUE #9387: Weird span for methods generated by proc macros
https://github.com/nox/proc-macro-fn-span/blob/2ae9a58221628f842a3719e425e3f3dce599c3d2/src/lib.rs#L14-L15

When hovering `.f1()`, rust-analyzer thinks the definition is on `struct`:

Capture d’écran 2021-06-23 à 15 08 32

When hovering `.f2()`, rust-analyzer thinks the definition is on the indentation spaces on the same line as the `f1` field:

Capture d’écran 2021-06-23 à 15 09 10

I'm pretty sure I set the spans correctly, as src links generated by rustdoc do point to the fields as expected.

This is on the latest rust-analyzer nightly, with `rust-analyzer.experimental.procAttrMacros` set to `true`.

### PR DESCRIPTION
**Pull request by Veykril**
Merely detaching the attributes causes incorrect spans to appear when mapping tokens up as the token ids resolve to the ranges of the stripped item so all the text ranges of its tokens are actually lower than the non-stripped ones.

Same fix as with attributes can be applied here, just replace the derive attribute with an equal amount of whitespace.

Fixes #9387

### CODE REVIEW

File: crates/hir_expand/src/input.rs [hunk 1]
```diff
@@ -95,13 +101,20 @@ struct A {
...
+                Token(WHITESPACE@33..49 "                ")
+                Token(WHITESPACE@49..50 "\n")
+                Node(ATTR@50..66)
+                Token(WHITESPACE@66..67 "\n")
+                Token(STRUCT_KW@67..73 "struct")
+                Token(WHITESPACE@73..74 " ")
+                Node(NAME@74..75)
+                Token(WHITESPACE@75..76 " ")
+                Node(RECORD_FIELD_LIST@76..92)
+            "#]],
```
Comment (Veykril) [inline crates/hir_expand/src/input.rs:117]: `expect!` doesn't work too well with whitespace lines so

**Comment by lnicola** [comment]
This increased the memory usage on `self` from 1087 to 1244 (!) MB.

**Comment by Veykril** [comment]
Technically expected because we were basically creating empty macro files for built-in attributes where as now we expand to the underlying item instead of discarding the annotated item.

So I guess this means we are duplicating all `#[test]` functions now into a macro file each, this does sound kinda unnecessary 🤔

I suppose we might just not wanna expand the `#[test]` attribute(or rather all builtins that expand to the item itself)?

**Comment by Veykril** [comment]
That seems very odd 😕

**Pull request merged**